src/core/internal/hash.d

```diff
@@ -30,6 +30,134 @@ private enum floatCoalesceNaNs = true;
+        enum canBitwiseHash = false;
+    else static if (is(T == struct))
+    {
+        static if (hasCallableToHash!T || __traits(isNested, T))
+            enum canBitwiseHash = false;
+        else
+            enum canBitwiseHash = allSatisfy!(.canBitwiseHash, typeof(T.tupleof));
+    }
+    else static if (is(T == union))
+    {
+        static if (hasCallableToHash!T)
+            enum canBitwiseHash = false;
+        else
+            // Right now we always bytewise hash unions
+            // that lack a callable `toHash`.
+            enum canBitwiseHash = true;
+    }
+    else static if (__traits(isAssociativeArray, T))
+    {
+        enum canBitwiseHash = false;
+    }
+    else static if (is(T E : E[]))
+    {
+        static if (__traits(isStaticArray, T))
+            enum canBitwiseHash = (T.length == 0) || .canBitwiseHash!E;
+        else
+            enum canBitwiseHash = false;
+    }
+    else static if (__traits(isFloating, T))
+    {
+        enum canBitwiseHash = !(floatCoalesceZeroes || floatCoalesceNaNs);
+    }
+    else
+    {
+        static assert(__traits(isScalar, T) || is(T == delegate) ||
+            is(T V : V*) || is(T : typeof(null)),
+            "Internal error: unanticipated type "~T.tostring);
+        enum canBitwiseHash = true;
+    }
+}
+
+private template UnqualUnsigned(T) if (__traits(isIntegral, T))
+{
+    static if (T.sizeof == ubyte.sizeof) alias UnqualUnsigned = ubyte;
+    else static if (T.sizeof == ushort.sizeof) alias UnqualUnsigned = ushort;
+    else static if (T.sizeof == uint.sizeof) alias UnqualUnsigned = uint;
+    else static if (T.sizeof == ulong.sizeof) alias UnqualUnsigned = ulong;
+    else static if (T.sizeof == ulong.sizeof * 2)
+    {
+        static assert(T.sizeof == ucent.sizeof);
+        alias UnqualUnsigned = ucent;
+    }
+    else
+    {
+        static assert(0, "No known unsigned equivalent of " ~ T.stringof);
+    }
+
+    static assert(UnqualUnsigned.sizeof == T.sizeof && __traits(isUnsigned, UnqualUnsigned));
+}
+
+//enum hash. CTFE depends on base type
+size_t hashOf(T)(scope const T val, size_t seed = 0)
+if (is(T == enum) && (__traits(isScalar, T) ||
+    (is(T : E[], E) && __traits(isScalar, E) && !__traits(isStaticArray, T)
+        && !is(T == class) && !is(T == struct) && !is(T == union))))
+{
+    static if (is(T EType == enum)) //for EType
+    {
+        return hashOf(cast(const EType) val, seed);
+    }
+    else
+    {
+        static assert(0);
+    }
+}
+
+//enum hash. CTFE depends on base type
+size_t hashOf(T)(auto ref T val, size_t seed = 0)
+if (is(T == enum) && !(__traits(isScalar, T) ||
+    (is(T : E[], E) && __traits(isScalar, E) && !__traits(isStaticArray, T)
+        && !is(T == class) && !is(T == struct) && !is(T == union))))
+{
+    static if (is(T EType == enum)) //for EType
+    {
+        EType e_val = cast(EType)val;
+        return hashOf(e_val, seed);
+    }
+    else
+    {
+        static assert(0);
+    }
+}
+
```

Comment:
The function body is equivalent to the hashOf below :/

Reply:
Fixed.